edb/edgeql/parser/parser.py

```diff
@@ -36,9 +37,64 @@ def get_exception(self, native_err, context, token=None):
         else:
             if msg.startswith('Unexpected token: '):
                 token = token or getattr(native_err, 'token', None)
+                ltok = self.parser._stack[-1][0]
 
                 if not token or token.kind() == 'EOF':
                     msg = 'Unexpected end of line'
+                elif (
+                    token.kind() == 'IDENT' and
+                    isinstance(ltok, parsing.Nonterm)
+                ):
+                    # Make sure that the previous element in the stack
+                    # is some kind of Nonterminal, because if it's
+                    # not, this is probably not an issue of a missing
+                    # COMMA.
+
+                    msg = f'Unexpected {token.text()!r}'
+                    # There are a few things this could be, so we make
+                    # an attempt to distinguish some specific parsing
+                    # error in shapes.
+
+                    # Then check if LBRACE, LPAREN, or LBRACKET is in
+                    # the parser stack. Any one of those indicates
+                    # that the parser was in the middle of processing
+                    # a list of some sort (shape, tuple, or array) and
+                    # encountering an unexpected IDENT probably
+                    # indicates a missing COMMA instead.
+                    for i, (el, _) in enumerate(reversed(self.parser._stack)):
+                        if isinstance(el, tokens.Token):
+                            if isinstance(el, tokens.T_LBRACE):
+                                if (
+                                    isinstance(ltok, gr_exprs.Identifier) and
+                                    ltok.val in keywords.edgeql_keywords
+                                ):
+                                    # This is some [unreserved]
+                                    # keyword preceding the unexpected
+                                    # IDENT in a shape, which probably
+                                    # is not a case of a missing COMMA
+                                    # anymore.
+                                    break
+                                else:
+                                    msg = (f"Missing ',' before the shape "
```

Comment:
What about `SELECT {1 2}`? It's not a shape.

Reply:
You're right. I was just focusing on the unexpected `IDENT` errors, but I can reframe it so that we check for this, too, by swapping the order of ifs and looking at the parsing stack first to determine the context and then narrowing it down by considering the unexpected token itself. Since this is in the error path, the tiny overhead of looping through the stack first shouldn't matter.